src/gsd.js

```diff
@@ -1,177 +1,223 @@
-        ddY = SG(y, x[1]-x[0], {windowSize:sgOptions.windowSize, polynomial:sgOptions.polynomial,derivative:2});
-    }
-    else {
-        if(options.smoothY)
-            Y = SG(y, x, {windowSize:sgOptions.windowSize, polynomial:sgOptions.polynomial,derivative:0});
-        dY = SG(y, x, {windowSize:sgOptions.windowSize, polynomial:sgOptions.polynomial,derivative:1});
-        ddY = SG(y, x, {windowSize:sgOptions.windowSize, polynomial:sgOptions.polynomial,derivative:2});
+    // If the max difference between delta x is less than 5%, then, we can assume it to be equally spaced variable
+    let Y = y;
+    let dY, ddY;
+    if ((maxDx - minDx) / maxDx < 0.05) {
+        if (options.smoothY)
+            Y = SG(y, x[1] - x[0], {windowSize: sgOptions.windowSize, polynomial: sgOptions.polynomial, derivative: 0});
+        dY = SG(y, x[1] - x[0], {windowSize: sgOptions.windowSize, polynomial: sgOptions.polynomial, derivative: 1});
+        ddY = SG(y, x[1] - x[0], {windowSize: sgOptions.windowSize, polynomial: sgOptions.polynomial, derivative: 2});
+    } else {
+        if (options.smoothY)
+            Y = SG(y, x, {windowSize: sgOptions.windowSize, polynomial: sgOptions.polynomial, derivative: 0});
+        dY = SG(y, x, {windowSize: sgOptions.windowSize, polynomial: sgOptions.polynomial, derivative: 1});
+        ddY = SG(y, x, {windowSize: sgOptions.windowSize, polynomial: sgOptions.polynomial, derivative: 2});
     }
 
     const X = x;
     const dx = x[1] - x[0];
     var maxDdy = 0;
     var maxY = 0;
-    for (i = 0; i < Y.length ; i++) {
-        if(Math.abs(ddY[i]) > maxDdy) {
+    for (let i = 0; i < Y.length; i++) {
+        if (Math.abs(ddY[i]) > maxDdy) {
             maxDdy = Math.abs(ddY[i]);
         }
-        if(Math.abs(Y[i]) > maxY) {
+        if (Math.abs(Y[i]) > maxY) {
             maxY = Math.abs(Y[i]);
         }
     }
 
-    var minddY = [];
-    var intervalL = [];
-    var intervalR = [];
     var lastMax = null;
     var lastMin = null;
-    var broadMask = new Array();
-    //By the intermediate value theorem We cannot find 2 consecutive maxima or minima
-    for (i = 1; i < Y.length -1 ; i++){
-        //console.log(dY[i]);
-        if ((dY[i] < dY[i-1]) && (dY[i] <= dY[i+1]) ||
-            (dY[i] <= dY[i-1]) && (dY[i] < dY[i+1])) {
-            lastMin = X[i];
-            if(dx > 0 && lastMax != null) {
-                intervalL.push(lastMax);
-                intervalR.push(lastMin);
+    var minddY = new Array(Y.length - 2);
+    var intervalL = new Array(Y.length);
+    var intervalR = new Array(Y.length);
+    var broadMask = new Array(Y.length - 2);
+    var minddYLen = 0;
+    var intervalLLen = 0;
+    var intervalRLen = 0;
+    var broadMaskLen = 0;
+    // By the intermediate value theorem We cannot find 2 consecutive maximum or minimum
+    for (let i = 1; i < Y.length - 1; ++i) {
 
+        // filter based on derivativeThreshold
+        if (Math.abs(dY[i]) > options.derivativeThreshold) {
```

Comment:
cc @andcastillo the comment it's not updated, but here is how the code is now, but I think that should be like this, not like `if (Math.abs(dY[i]) < options.derivativeThreshold) {`